### PR TITLE
feat: update global farm

### PR DESCRIPTION
**Pull request by dmoka**
Update global farm parameters

- [x] benchmark on reference
- [x] add invariant tests

### CODE REVIEW

File: integration-tests/src/omnipool_liquidity_mining.rs [hunk 1]
```diff
@@ -1238,6 +1238,134 @@ fn liquidity_mining_should_work_when_farm_distribute_bonds() {
...
+		));
+
+		let position_id = omnipool_add_liquidity(CHARLIE.into(), DOT, 1 * UNITS);
+		assert_nft_owner!(
+			hydradx_runtime::OmnipoolCollectionId::get(),
+			position_id,
+			CHARLIE.into()
+		);
+
+		set_relaychain_block_number(400);
```
Comment (martinfridrich) [inline integration-tests/src/omnipool_liquidity_mining.rs:1305]: wrong block number, you can't go back in time

File: pallets/liquidity-mining/src/lib.rs [hunk 2]
```diff
@@ -569,6 +569,59 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+			Error::<T, I>::InvalidPlannedYieldingPeriods
+		);
+		ensure!(!yield_per_period.is_zero(), Error::<T, I>::InvalidYieldPerPeriod);
+
+		<GlobalFarm<T, I>>::try_mutate(global_farm_id, |maybe_global_farm| {
+			let global_farm = maybe_global_farm.as_mut().ok_or(Error::<T, I>::GlobalFarmNotFound)?;
+
+			ensure!(global_farm.state.is_active(), Error::<T, I>::GlobalFarmNotFound);
+
+			//Sync global farm to get right accumulated_rpz and pending rewards
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/lib.rs:598]: I don't find this comment helpful, it basically says same thing as doc comment of `sync_global_farm`

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 3]
```diff
@@ -506,6 +506,20 @@ prop_compose! {
 	}
 }
 
+prop_compose! {
+	fn arb_deposit2()(
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:510]: Why did you create special deposit for `update_global_farm_invariant_1`  with different `valued_shares` range than other invariant tests?

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 4]
```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
...
+	//Index of deposit in `deposit` vec. This idx is used in each test case run and execute claim
+	//if deposit exits.
+
+	let deposit_idx_range = 0..500_usize;
+
+	invariants_externalities().execute_with(|| {
+		let mut runner = TestRunner::new(Config {
+			cases: successfull_cases,
+			source_file: Some("liquidity-mining/src/tests/invariants.rs"),
+			test_name: Some("update_global_farm_invariant_first"),
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:936]:
```suggestion
			test_name: Some("update_global_farm_invariant_1"),
```

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 5]
```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
...
+							d.global_farm_id,
+							planned_yielding_period,
+							yield_per_period,
+							1_000,
+						)
+						.unwrap();
+						TransactionOutcome::Commit(DispatchResult::Ok(()))
+					});
+					let g_farm = LiquidityMining::global_farm(d.global_farm_id).unwrap();
+					distributed_before_update.borrow_mut()[d.global_farm_id as usize] =
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:969]:
```suggestion
//NOTE: `update_global_farm()` distributes rewards before update so this must be done after function execution but before any interaction with farm.
					distributed_before_update.borrow_mut()[d.global_farm_id as usize] =
```
or something similar so it's clear why it's named  "before update" but value is taken after farm's update

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 6]
```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
...
+							assert!(
+								gf.total_rewards >= s_1,
+								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
+							);
+							assert_eq_approx!(
+								gf.total_rewards,
+								s_1,
+								10_000_000,
+								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
+							);
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:1022]:
```suggestion
							assert!(
								gf.total_rewards >= s_1,
								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
							);
							//NOTE: Approax becasue of div in max_reward_per_period calculation.
							assert_eq_approx!(
								gf.total_rewards,
								s_1,
								10_000_000,
								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
							);
```

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 7]
```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
...
+			)
+			.unwrap();
+	});
+}
+
+#[test]
+//Update global farm for invariants 1,2,3
+//https://www.notion.so/Liquidity-mining-spec-b30ccfe470a74173b82c3702b1e8fca1#422dea2e23744859baeb704dbdb3caca
+//
+fn update_global_farm_invariant_1_2_3() {
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:1037]: invariant 1 has its' own test so this name is not correct, please fix it also in comment and `test_name`

File: pallets/liquidity-mining/src/tests/invariants.rs [hunk 8]
```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
...
+						let claimed_amount = d.yield_farm_entries[0].accumulated_claimed_rewards;
+						claimed_by_users_sum += claimed_amount;
+					}
+
+					//WARN: There is no room for rounding errors in this invariant. Any discrepancy
+					//in this assert means we are loosing tokens somewhere.
+					assert_eq!(
+						total_rewards_sum,
+						farm_balances_sum + pot_balance_sum + claimed_by_users_sum,
+					);
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/invariants.rs:1190]: This is not invariant 3 check, this is same check as invariant 2 block above

File: pallets/liquidity-mining/src/tests/update_global_farm.rs [hunk 9]
```diff
@@ -182,3 +183,124 @@ fn update_global_farm_price_adjustment_should_fail_when_farm_is_already_terminat
...
+					zero_yield_per_period,
+					MIN_DEPOSIT,
+				),
+				Error::<Test, Instance1>::InvalidYieldPerPeriod
+			);
+
+			TransactionOutcome::Commit(DispatchResult::Ok(()))
+		});
+	});
+}
```
Comment (martinfridrich) [inline pallets/liquidity-mining/src/tests/update_global_farm.rs:306]: I would like to also see test where user claimed correct amount after farms update, something like:
* run for few periods
* update global farm
* run for few periods
* claim assert for correct amount

File: pallets/liquidity-mining/src/tests/invariants.rs (hunk 3 above)
Comment (dmoka) [reply pallets/liquidity-mining/src/tests/invariants.rs:510]: I had some strange failing tests, namely IncorrectValueShares, but now changing back it looks good, so we can get rid of this.

File: pallets/liquidity-mining/src/tests/invariants.rs (hunk 8 above)
Comment (dmoka) [reply pallets/liquidity-mining/src/tests/invariants.rs:1177]: damn, good catch, fixed!

File: pallets/liquidity-mining/src/tests/update_global_farm.rs (hunk 9 above)
Comment (dmoka) [reply pallets/liquidity-mining/src/tests/update_global_farm.rs:306]: added, pls check it!

**Review by martinfridrich — APPROVED** [review]

**Review by jak-pan — APPROVED** [review]

**Pull request merged**